external/storm-hbase/src/main/java/org/apache/storm/hbase/state/HBaseKeyValueState.java

```diff
@@ -0,0 +1,399 @@
+        try {
+            Result[] results = hBaseClient.batchGet(Collections.singletonList(get));
+            Result result = results[0];
+            if (!result.isEmpty()) {
+                LOG.debug("Loading previously prepared commit from {}", prepareNamespace);
+                NavigableMap<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
+                pendingCommit = ByteArrayUtil.Maps.newHashMapWrappingKey(familyMap);
+            } else {
+                LOG.debug("No previously prepared commits.");
+                pendingCommit = Collections.emptyMap();
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void put(K key, V value) {
+        LOG.debug("put key '{}', value '{}'", key, value);
+        byte[] columnKey = encoder.encodeKey(key);
+        byte[] columnValue = encoder.encodeValue(value);
+        pendingPrepare.put(new ByteArrayWrapper(columnKey), columnValue);
+    }
+
+    @Override
+    public V get(K key) {
+        LOG.debug("get key '{}'", key);
+        byte[] columnKey = encoder.encodeKey(key);
+        byte[] columnValue = null;
+
+        ByteArrayWrapper wrappedKey = new ByteArrayWrapper(columnKey);
+        if (pendingPrepare.containsKey(wrappedKey)) {
+            columnValue = pendingPrepare.get(wrappedKey);
+        } else if (pendingCommit.containsKey(wrappedKey)) {
+            columnValue = pendingCommit.get(wrappedKey);
+        } else {
+            HBaseProjectionCriteria criteria = new HBaseProjectionCriteria();
+            HBaseProjectionCriteria.ColumnMetaData column = new HBaseProjectionCriteria.ColumnMetaData(columnFamily, columnKey);
+            criteria.addColumn(column);
+            Get get = hBaseClient.constructGetRequests(namespace, criteria);
+            try {
+                Result[] results = hBaseClient.batchGet(Collections.singletonList(get));
+                Result result = results[0];
+                columnValue = result.getValue(column.getColumnFamily(), column.getQualifier());
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+        V value = null;
+        if (columnValue != null) {
+            value = encoder.decodeValue(columnValue);
+        }
+        LOG.debug("Value for key '{}' is '{}'", key, value);
+        return value;
+    }
+
+    @Override
+    public V get(K key, V defaultValue) {
+        V val = get(key);
+        return val != null ? val : defaultValue;
+    }
+
+    @Override
+    public V delete(K key) {
+        LOG.debug("delete key '{}'", key);
+        byte[] columnKey = encoder.encodeKey(key);
+        V curr = get(key);
+        pendingPrepare.put(new ByteArrayWrapper(columnKey), HBaseEncoder.TOMBSTONE);
+        return curr;
+    }
+
+    @Override
+    public Iterator<Map.Entry<K, V>> iterator() {
+        // TODO: is it possible to implement this correctly?
```

Comment:
@HeartSaVioR nice observation. Is it because the entries in the pending iterators can get committed to the db (redis) while we are iterating ? If so we may be we need to fix it. While returning the values from the pending iterators we could store the keys into a set and use this to filter already returned keys while scanning the db (redis).

We put 'tombstone' entries into the pending maps during delete, but I don't see this handled during the iteration. May be we need to handle this as well.

Once the redis implementation is fixed, may be you can refactor and re-use the common parts for hbase as well.